official/resnet/ctl/ctl_imagenet_main.py

```diff
@@ -0,0 +1,297 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Runs a ResNet model on the ImageNet dataset using custom training loops."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import tempfile
+import datetime
+import time
+
+from absl import app as absl_app
+from absl import flags
+# from absl import logging
+
+import tensorflow.compat.v1.logging as logging
+import tensorflow as tf
+import numpy as np
+
+from official.resnet import imagenet_main
+from official.resnet.keras import keras_common
+from official.resnet.keras import resnet_model
+from official.utils.flags import core as flags_core
+from official.utils.logs import logger
+from official.utils.misc import distribution_utils
+from official.utils.misc import model_helpers
+from official.resnet.ctl import ctl_common
+
+
+# Imagenet training and test data sets.
+APPROX_IMAGENET_TRAINING_IMAGES = 1280000  # Approximate number of images.
+IMAGENET_VALIDATION_IMAGES = 50000  # Number of images.
+
+DEFAULT_IMAGE_SIZE = 224
+NUM_CHANNELS = 3
+NUM_CLASSES = 1001
+
+
+def parse_record_keras(raw_record, is_training, dtype):
+  """Adjust the shape of label."""
+  image, label = imagenet_main.parse_record(raw_record, is_training, dtype)
+
+  # Subtract one so that labels are in [0, 1000), and cast to float32 for
+  # Keras model.
+  label = tf.cast(tf.cast(tf.reshape(label, shape=[1]), dtype=tf.int32) - 1,
+                  dtype=tf.float32)
+  return image, label
+
+# Learning rate schedule
+_LR_SCHEDULE = [    # (multiplier, epoch to start) tuples
+    (1.0, 5), (0.1, 30), (0.01, 60), (0.001, 80)
+]
+_BASE_LEARNING_RATE = 0.4
+
+
+def compute_learning_rate(lr_epoch):
+  """Learning rate for each step."""
+  warmup_lr_multiplier, warmup_end_epoch = _LR_SCHEDULE[0]
+  if lr_epoch < warmup_end_epoch:
+    # Learning rate increases linearly per step.
+    return (_BASE_LEARNING_RATE * warmup_lr_multiplier *
+            lr_epoch / warmup_end_epoch)
+  for mult, start_epoch in _LR_SCHEDULE:
+    if lr_epoch >= start_epoch:
+      learning_rate = _BASE_LEARNING_RATE * mult
+    else:
+      break
+  return learning_rate
+
+
+def run(flags_obj):
```

Comment:
This function seems massive with all the inline functions.  Is this how we would suggest people do it?

Reply:
Moved a few things around. Let me know if this helps simplify things a bit.